src/btree/bt_page.c

```diff
@@ -418,20 +454,178 @@ __wt_page_inmem(WT_SESSION_IMPL *session, WT_REF *ref, const void *image, uint32
     return (ret);
 }
 
+/*
+ * __wt_col_fix_read_auxheader --
+ *     Read the auxiliary header following the bitmap data, if any. This code is used by verify and
+ *     needs to be accordingly careful. It is also used by mainline reads so it must also not crash
+ *     or print.
+ */
+int
+__wt_col_fix_read_auxheader(
+  WT_SESSION_IMPL *session, const WT_PAGE_HEADER *dsk, WT_COL_FIX_AUXILIARY_HEADER *auxhdr)
+{
+    WT_BTREE *btree;
+    uint32_t bitmapsize, auxheaderoffset;
+    const uint8_t *raw, *end;
+
+    btree = S2BT(session);
+
+    /*
+     * Figure where the auxiliary header is. It is always immediately after the bitmap data,
+     * regardless of whether the page is full.
+     */
+    bitmapsize = __bitstr_size(dsk->u.entries * btree->bitcnt);
+    auxheaderoffset = WT_PAGE_HEADER_BYTE_SIZE(btree) + bitmapsize;
+
+    /*
+     * If the auxiliary header is past the in-memory page size, there's no auxiliary data. If
+     * there's at least one byte past the bitmap data, check whether it's zero. If that's zero,
+     * there's no auxiliary data. (We are guaranteed that any allocation slop that we might be
+     * looking at is all zeros.) Set everything to zero and return.
+     */
+    if (auxheaderoffset >= dsk->mem_size || *(raw = (uint8_t *)dsk + auxheaderoffset) == 0) {
+        auxhdr->version = WT_COL_FIX_VERSION_NIL;
+        auxhdr->entries = 0;
+        auxhdr->offset = 0;
+        return (0);
+    }
+
+    /* Remember the end of the page for easy computation of maximum lengths. */
+    end = (uint8_t *)dsk + dsk->mem_size;
+
+    /*
+     * The on-disk header is a 1-byte version, a uint32 with the number of entries, and a uint32
+     * that says how many waste bytes there are between the header and the entries.
+     */
+    auxhdr->version = raw[0];
+    raw++;
+    {
+        uint32_t entries, waste;
+
+        if (raw + sizeof(entries) > end)
+            return (EINVAL);
+        memcpy(&entries, raw, sizeof(entries));
+        raw += sizeof(entries);
+        if (raw + sizeof(waste) > end)
+            return (EINVAL);
+        memcpy(&waste, raw, sizeof(waste));
+        raw += sizeof(waste);
+
+#ifdef WORDS_BIGENDIAN
+        entries = __wt_bswap32(entries);
+        waste = __wt_bswap32(waste);
+#endif
+
+        if (raw + waste > end)
+            return (EINVAL);
```

Comment:
I specifically didn't want to paste this function twice, but I guess I can pass in a flag that changes the checks to assertions. Knowingly reading off the end of a memory block really goes against the grain though, and sooner or later some static analyzer will carp about it, so I do want to leave assertions.

I have done this; it's in a second commit. Please take a look and see what you think...

Reply:
With respect to doing validity checking in the page-read code, I disagree pretty strongly on this one. These should be two functions, because one function should be checking all the edge cases it can find and the other should be using the information without  any checking at all. That's the way the rest of the engine works, and I think it's the right approach. We don't do validity checking when we read pages, that  leads to bad performance.

Reply:
There are three checks now; one can just be removed entirely, the second one can be moved outside the function, and the third is obsoleted by the pending header format change. (It's equivalent to and replaced by the bounds checks in __wt_vunpack_uint, which aren't optional. So I think this is a dead article and we still only need one function. (Or at least the disagreement is a dead article. Still need to rearrange the other two checks.)

Reply:
Fixed. Verify remains paranoid, debug prints messages if things look bad, inmem is like one cycle faster (but will plunge ahead if the page version is unexpected, instead of raising a panic), salvage has fewer redundant checks. And we don't need two copies of the function.